File: dungeon-crawler/android/src/main/java/com/novoda/dungeoncrawler/InitHack.java
package com.novoda.dungeoncrawler;

import com.novoda.dungeoncrawler.DungeonCrawlerGame.HudDisplayer;
import com.yheriatovych.reductor.Store;

public class InitHack {

    private static DungeonCrawlerGame game;
    private static final Store<Redux.GameState> store = Store.create(new Redux.GameReducer(), Redux.GameState.getInitialState(), new FirebaseGameStateLogger());

    public static DungeonCrawlerGame newInstance(int numOfSquares,
                                                 Display display,
                                                 HudDisplayer hudDisplayer,
                                                 JoystickActuator joystickActuator,
                                                 SoundEffectsPlayer soundEffectsPlayer,
                                                 Screensaver screensaver, ArduinoLoop looper) {
        StartClock clock = new PausableStartClock(new DefaultStartClock());
        new LoopObserver(
            store,
            joystickActuator,
            () -> game.onAttack(),
            v -> game.onMove(v),
            () -> game.onDeath(),
            () -> game.onKill()
        );
        new FrameObserver(
            store,
            screensaver::draw,
            (l, l2) -> game.onWin(l, l2),
            (s, c) -> game.onGameComplete(s, c),
            () -> game.onGameOver(),
            new FrameObserver.DrawCallback() {

                @Override
                public void drawPlayer(int position) {
                    game.drawPlayer(position);
                }

                @Override
                public void drawConveyor(int startPoint, int endPoint, Direction direction, long frame) {
                    game.drawConveyor(startPoint, endPoint, direction, frame);
                }

                @Override
                public void drawAttack(int startPoint, int centerPoint, int endPoint, int attackPower) {
                    game.drawAttack(startPoint, centerPoint, endPoint, attackPower);
                }

                @Override
                public void drawParticle(int position, int power) {
                    game.drawParticle(position, power);
                }

                @Override
                public void drawEnemy(int position) {
                    game.drawEnemy(position);
                }

                @Override
                public void drawLava(int startPosition, int endPosition, boolean enabled) {
                    game.drawLava(startPosition, endPosition, enabled);
                }

                @Override
                public void drawBoss(int startPosition, int endPosition) {
                    game.drawBoss(startPosition, endPosition);
                }

                @Override
                public void drawExit() {
                    game.drawExit();
                }

                @Override
                public void drawLives(int lives) {
                    game.drawLives(lives);
                }

            },
            new FrameObserver.FrameCallback() {
                @Override
                public void onFrameStart() {
                    game.onFrameStart();
                }

                @Override
                public void onFrameEnd() {
                    game.onFrameEnd();
                }
            },
            clock
        );
        new TimeObserver(store, clock);
        GameEngine gameEngine = new PauseResumeGameEngine(
            store,
            joystickActuator,
            clock
        );
        game = new DungeonCrawlerGame(
            numOfSquares,
            gameEngine,
            display,
            hudDisplayer,
            soundEffectsPlayer,
            looper
        );
        return game;
    }

}

File: dungeon-crawler/android_companion/src/main/java/com/novoda/dungeoncrawlercompanion/InitHack.kt
package com.novoda.dungeoncrawlercompanion

import com.novoda.dungeoncrawler.ArduinoLoop
import com.novoda.dungeoncrawler.DefaultStartClock
import com.novoda.dungeoncrawler.Direction
import com.novoda.dungeoncrawler.Display
import com.novoda.dungeoncrawler.DungeonCrawlerGame
import com.novoda.dungeoncrawler.DungeonCrawlerGame.HudDisplayer
import com.novoda.dungeoncrawler.FrameObserver
import com.novoda.dungeoncrawler.Redux
import com.yheriatovych.reductor.Store

object InitHack {

    private val store = Store.create(ReplayReducer(), Redux.GameState.getInitialState())

    fun newInstance(numOfSquares: Int,
                    display: Display,
                    hudDisplayer: HudDisplayer,
                    looper: ArduinoLoop): DungeonCrawlerGame {
        val clock = DefaultStartClock()
        val gameEngine = ReplayGameEngine(
                ReplayFetcher(),
                store,
                clock
        )
        val game = DungeonCrawlerGame(
                numOfSquares,
                gameEngine,
                display,
                hudDisplayer,
                NoOpSoundEffectsPlayer(),
                looper
        )

        FrameObserver(
                store,
                FrameObserver.NoInputMonitor { },
                FrameObserver.WinMonitor { l, l2 -> Unit },
                FrameObserver.CompleteMonitor { s, c -> Unit },
                FrameObserver.GameOverMonitor { Unit },
                createDrawCallback(game),
                createFrameCallback(game),
                clock
        )

        return game
    }

    private fun createDrawCallback(game: DungeonCrawlerGame): FrameObserver.DrawCallback {
        return object : FrameObserver.DrawCallback {

            override fun drawPlayer(position: Int) {
                game.drawPlayer(position)
            }

            override fun drawConveyor(startPoint: Int, endPoint: Int, direction: Direction, frame: Long) {
                game.drawConveyor(startPoint, endPoint, direction, frame)
            }

            override fun drawAttack(startPoint: Int, centerPoint: Int, endPoint: Int, attackPower: Int) {
                game.drawAttack(startPoint, centerPoint, endPoint, attackPower)
            }

            override fun drawParticle(position: Int, power: Int) {
                game.drawParticle(position, power)
            }

            override fun drawEnemy(position: Int) {
                game.drawEnemy(position)
            }

            override fun drawLava(startPosition: Int, endPosition: Int, enabled: Boolean) {
                game.drawLava(startPosition, endPosition, enabled)
            }

            override fun drawBoss(startPosition: Int, endPosition: Int) {
                game.drawBoss(startPosition, endPosition)
            }

            override fun drawExit() {
                game.drawExit()
            }

            override fun drawLives(lives: Int) {
                game.drawLives(lives)
            }

        }
    }

    private fun createFrameCallback(game: DungeonCrawlerGame): FrameObserver.FrameCallback {
        return object : FrameObserver.FrameCallback {
            override fun onFrameStart() {
                                        game.onFrameStart()
            }

            override fun onFrameEnd() {
                game.onFrameEnd()
            }
        }
    }

}
